// @ngx-lite/util/bundles/ngx-lite-util.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
    typeof define === 'function' && define.amd ? define('@ngx-lite/util', ['exports'], factory) :
    (factory((global['ngx-lite'] = global['ngx-lite'] || {}, global['ngx-lite'].util = {})));
}(this, (function (exports) { 'use strict';

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */
    /** @enum {number} */
    var KeyCodes = {
        LeftArrow: 37,
        UpArrow: 38,
        RightArrow: 39,
        DownArrow: 40,
        Backspace: 8,
        Tab: 9,
        Enter: 13,
        Escape: 27,
    };
    KeyCodes[KeyCodes.LeftArrow] = 'LeftArrow';
    KeyCodes[KeyCodes.UpArrow] = 'UpArrow';
    KeyCodes[KeyCodes.RightArrow] = 'RightArrow';
    KeyCodes[KeyCodes.DownArrow] = 'DownArrow';
    KeyCodes[KeyCodes.Backspace] = 'Backspace';
    KeyCodes[KeyCodes.Tab] = 'Tab';
    KeyCodes[KeyCodes.Enter] = 'Enter';
    KeyCodes[KeyCodes.Escape] = 'Escape';
    /**
     * @param {?} elm
     * @return {?}
     */
    function trapTabFocus(elm) {
        /** @type {?} */
        var focusableEls = elm.querySelectorAll('a, object, input, button, iframe, [tabindex]');
        /** @type {?} */
        var firstFocusableEl = /** @type {?} */ (focusableEls[0]);
        /** @type {?} */
        var lastFocusableEl = /** @type {?} */ (focusableEls[focusableEls.length - 1]);
        // if just a single focusable item focus it
        if (firstFocusableEl) {
            firstFocusableEl.focus();
        }
        // if two or more focusable items create focus loop
        if (firstFocusableEl && lastFocusableEl) {
            elm.addEventListener('keydown', function (e) {
                /** @type {?} */
                var isTabPressed = e.key === 'Tab' || e.keyCode === KeyCodes.Tab;
                if (!isTabPressed) {
                    return;
                }
                if (e.shiftKey && document.activeElement === firstFocusableEl) {
                    lastFocusableEl.focus();
                    e.preventDefault();
                }
                else {
                    if (document.activeElement === lastFocusableEl) {
                        firstFocusableEl.focus();
                        e.preventDefault();
                    }
                }
            });
        }
    }
    /**
     * @return {?}
     */
    function isBrowser() {
        return typeof window !== 'undefined';
    }
    /**
     * @return {?}
     */
    function lockScroll() {
        if (isBrowser()) {
            document.body.style.overflow = 'hidden';
        }
    }
    /**
     * @return {?}
     */
    function unlockScroll() {
        if (isBrowser()) {
            document.body.style.overflow = 'initial';
        }
    }
    /**
     * @return {?}
     */
    function ariaHideBody() {
        if (isBrowser()) {
            document.body.setAttribute('aria-hidden', 'true');
        }
    }
    /**
     * @return {?}
     */
    function ariaShowBody() {
        if (isBrowser()) {
            document.body.setAttribute('aria-hidden', 'false');
        }
    }

    /**
     * @fileoverview added by tsickle
     * @suppress {checkTypes,extraRequire,uselessCode} checked by tsc
     */

    exports.trapTabFocus = trapTabFocus;
    exports.isBrowser = isBrowser;
    exports.lockScroll = lockScroll;
    exports.unlockScroll = unlockScroll;
    exports.ariaHideBody = ariaHideBody;
    exports.ariaShowBody = ariaShowBody;
    exports.KeyCodes = KeyCodes;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
